src/Workspaces/Core/Portable/Remote/RemoteHostAssetSerialization.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Serialization;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal static class RemoteHostAssetSerialization
+    {
+        public static async Task WriteDataAsync(ObjectWriter writer, IRemotableDataService remotableDataService, int scopeId, Checksum[] checksums, CancellationToken cancellationToken)
+        {
+            writer.WriteInt32(scopeId);
+
+            // special case
+            if (checksums.Length == 0)
+            {
+                writer.WriteInt32(0);
+                return;
+            }
+
+            if (checksums.Length == 1)
+            {
+                var checksum = checksums[0];
+                var remotableData = (await remotableDataService.GetRemotableDataAsync(scopeId, checksum, cancellationToken).ConfigureAwait(false)) ?? RemotableData.Null;
```

Comment:
when does `GetRemotableDataAsync` retrun null?

Reply:
Not sure. http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Execution/AssetStorages.cs,145